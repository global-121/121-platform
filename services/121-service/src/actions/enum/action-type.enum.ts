export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
