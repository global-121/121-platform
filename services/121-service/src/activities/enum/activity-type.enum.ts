export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
