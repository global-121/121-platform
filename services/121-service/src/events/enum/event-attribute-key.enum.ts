export enum EventAttributeKeyEnum {
  reason = 'reason',
  oldValue = 'oldValue',
  newValue = 'newValue',
  fieldName = 'fieldName',
  duplicateWithRegistrationId = 'duplicateWithRegistrationId',
  duplicateWithRegistrationProgramId = 'duplicateWithRegistrationProgramId',
}
