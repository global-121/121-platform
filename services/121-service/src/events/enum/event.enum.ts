export enum EventEnum {
  registrationDataChange = 'registrationDataChange',
  financialServiceProviderChange = 'financialServiceProviderChange',
  registrationStatusChange = 'registrationStatusChange',
  ignoredDuplicate = 'ignoredDuplicate',
}
