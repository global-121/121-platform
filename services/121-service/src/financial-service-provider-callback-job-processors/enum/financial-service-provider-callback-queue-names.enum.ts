export enum FinancialServiceProviderCallbackQueueNames {
  safaricomTransferCallback = 'safaricomTransferCallback',
  safaricomTimeoutCallback = 'safaricomTimeoutCallback',
}
