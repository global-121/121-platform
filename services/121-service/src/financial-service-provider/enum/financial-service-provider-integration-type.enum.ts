export enum FinancialServiceProviderIntegrationType {
  api = 'api',
  csv = 'csv',
  xml = 'xml',
}
