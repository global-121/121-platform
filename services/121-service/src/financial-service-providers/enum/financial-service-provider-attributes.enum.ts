export enum FinancialServiceProviderAttributes {
  phoneNumber = 'phoneNumber',
  nationalId = 'nationalId',
  fullName = 'fullName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
  bankAccountNumber = 'bankAccountNumber',
  whatsappPhoneNumber = 'whatsappPhoneNumber',
}
