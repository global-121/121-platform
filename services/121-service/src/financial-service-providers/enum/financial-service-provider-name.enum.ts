export enum FinancialServiceProviderName {
  intersolveVoucherWhatsapp = 'Intersolve-voucher-whatsapp',
  intersolveVoucherPaper = 'Intersolve-voucher-paper',
  intersolveVisa = 'Intersolve-visa',
  intersolveJumboPhysical = 'Intersolve-jumbo-physical',
  africasTalking = 'Africas-talking',
  belcash = 'BelCash',
  vodacash = 'VodaCash',
  bobFinance = 'BoB-finance',
  ukrPoshta = 'UkrPoshta',
  safaricom = 'Safaricom',
  commercialBankEthiopia = 'Commercial-bank-ethiopia',
  excel = 'Excel',
}

export enum FinancialServiceProviderConfigurationEnum {
  password = 'password',
  username = 'username',
  columnsToExport = 'columnsToExport',
  columnToMatch = 'columnToMatch',
  brandCode = 'brandCode',
  displayName = 'displayName',
  coverLetterCode = 'coverLetterCode',
}

export const FinancialServiceProviderConfigurationMapping: {
  [key in FinancialServiceProviderName]?: any;
} = {
  [FinancialServiceProviderName.intersolveVoucherWhatsapp]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVoucherPaper]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVisa]: [
    FinancialServiceProviderConfigurationEnum.brandCode,
    FinancialServiceProviderConfigurationEnum.coverLetterCode,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveJumboPhysical]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.africasTalking]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.belcash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.vodacash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.bobFinance]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.ukrPoshta]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.safaricom]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.commercialBankEthiopia]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.excel]: [
    FinancialServiceProviderConfigurationEnum.columnsToExport,
    FinancialServiceProviderConfigurationEnum.columnToMatch,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
};
