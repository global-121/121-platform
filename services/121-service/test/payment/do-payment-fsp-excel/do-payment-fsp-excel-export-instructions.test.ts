/* eslint-disable jest/no-conditional-expect */
import { HttpStatus } from '@nestjs/common';

import {
  FspConfigurationProperties,
  Fsps,
} from '@121-service/src/fsps/enums/fsp-name.enum';
import { TransactionStatusEnum } from '@121-service/src/payments/transactions/enums/transaction-status.enum';
import { GenericRegistrationAttributes } from '@121-service/src/registration/enum/registration-attribute.enum';
import { SeedScript } from '@121-service/src/scripts/enum/seed-script.enum';
import programTest from '@121-service/src/seed-data/program/program-test.json';
import {
  getFspInstructions,
  getTransactions,
} from '@121-service/test/helpers/program.helper';
import {
  deleteProgramFspConfigurationProperty,
  getProgramFspConfigurations,
} from '@121-service/test/helpers/program-fsp-configuration.helper';
import { seedPaidRegistrations } from '@121-service/test/helpers/registration.helper';
import {
  getAccessToken,
  resetDB,
} from '@121-service/test/helpers/utility.helper';
import {
  programIdWesteros,
  registrationCbe,
  registrationWesteros1,
  registrationWesteros2,
  registrationWesteros3,
} from '@121-service/test/registrations/pagination/pagination-data';

describe('Do payment with Excel FSP', () => {
  let accessToken: string;
  // Payment info
  const amount = 10;
  // Registrations
  const registrationsWesteros = [
    registrationWesteros1,
    registrationWesteros2,
    registrationWesteros3,
  ];

  const registrationsProgramWithValidation = [registrationCbe];

  const programIdCbe = 1;

  let paymentIdWesteros: number;
  let voucherPaymentIdWesteros: number;
  let paymentIdCbe: number;

  const gringotts = 'gringotts';
  const ironBank = 'ironBank';

  const columnsToExportGringotts = (programTest.programFspConfigurations
    .find((p) => p.name === gringotts)!
    .properties.find(
      (prop) => prop.name === FspConfigurationProperties.columnsToExport,
    )?.value ?? []) as string[];
  const columnsToExportIronBank = (programTest.programFspConfigurations
    .find((p) => p.name === ironBank)!
    .properties.find(
      (prop) => prop.name === FspConfigurationProperties.columnsToExport,
    )?.value ?? []) as string[];

  // Function for all seeding
  const seedPrograms = async () => {
    await resetDB(SeedScript.testMultiple, __filename);
    accessToken = await getAccessToken();

    //////////////////////////
    // Setup Westeros program
    /////////////////////////

    paymentIdWesteros = await seedPaidRegistrations(
      registrationsWesteros,
      programIdWesteros,
      amount,
      [TransactionStatusEnum.waiting],
    );

    // also do another payment with a different fsp to test export where there is not transaction with excel fsp
    const voucherRegistrationWesteros = {
      ...registrationWesteros1,
      referenceId: 'AH-001',
      phoneNumber: '31612345678',
      programFspConfigurationName: Fsps.intersolveVoucherWhatsapp,
    };

    voucherPaymentIdWesteros = await seedPaidRegistrations(
      [voucherRegistrationWesteros],
      programIdWesteros,
      amount,
      [TransactionStatusEnum.waiting],
    );

    ////////////////////////////
    // Setup Validation program
    ////////////////////////////

    // Do more tests with multiple programs, to include data isolation in tests
    // Specifically, this enables testing if transactions and registrations have the same length (see excel.service.ts)
    paymentIdCbe = await seedPaidRegistrations(
      registrationsProgramWithValidation,
      programIdCbe,
      amount,
      [TransactionStatusEnum.success],
    );
  };

  describe('Export FSP instructions successfully', () => {
    beforeEach(async () => {
      await seedPrograms();
    });
    it('Should return specified columns on Get FSP instruction with Excel-FSP when "columnsToExport" is set', async () => {
      // Arrange

      // Act
      const transactionsResponse = await getTransactions({
        programId: programIdWesteros,
        paymentId: paymentIdWesteros,
        registrationReferenceId: null,
        accessToken,
      });

      const fspInstructionsResponse = await getFspInstructions(
        programIdWesteros,
        paymentIdWesteros,
        accessToken,
      );
      const fspInstructions = fspInstructionsResponse.body;

      // Assert
      for (const transaction of transactionsResponse.body) {
        expect(transaction.status).toBe(TransactionStatusEnum.waiting);
      }
      // Sort fspInstructions by phoneNumber
      expect(fspInstructionsResponse.statusCode).toBe(HttpStatus.OK);

      // Find each instructions by fsp configuration name
      const fspInstructionsGringotts = fspInstructions.find(
        (f) => f.fileNamePrefix === gringotts,
      );
      const fspInstructionsIronBank = fspInstructions.find(
        (f) => f.fileNamePrefix === ironBank,
      );
      expect(fspInstructionsGringotts).toBeDefined();
      expect(fspInstructionsIronBank).toBeDefined();

      // Check if instructions contain the right data
      const gringottsData = fspInstructionsGringotts.data;
      const ironBankData = fspInstructionsIronBank.data;
      expect(gringottsData.length).toBe(1);
      expect(ironBankData.length).toBe(2);

      // Validate data per reference id
      const createFspInstructionObject = (
        registration: Record<string, unknown>,
        columnsToExport: string[],
      ): Record<string, unknown> => {
        // Any number for amount
        const obj: Record<string, unknown> = {
          amount: expect.any(Number), // Amount can differ per registration due payment amount calculation which is tested in different tests
          referenceId: registration.referenceId,
        };
        for (const column of columnsToExport) {
          obj[column] =
            registration[column] !== undefined
              ? String(registration[column])
              : '';
        }
        return obj;
      };

      const fspInstructionReg1 = ironBankData.find(
        (d) => d.referenceId === registrationWesteros1.referenceId,
      );
      expect(fspInstructionReg1).toEqual(
        createFspInstructionObject(
          registrationWesteros1,
          columnsToExportIronBank,
        ),
      );

      const fspInstructionReg2 = ironBankData.find(
        (d) => d.referenceId === registrationWesteros2.referenceId,
      );
      expect(fspInstructionReg2).toEqual(
        createFspInstructionObject(
          registrationWesteros2,
          columnsToExportIronBank,
        ),
      );

      const fspInstructionReg3 = gringottsData.find(
        (d) => d.referenceId === registrationWesteros3.referenceId,
      );
      expect(fspInstructionReg3).toEqual(
        createFspInstructionObject(
          registrationWesteros3,
          columnsToExportGringotts,
        ),
      );
    });

    it('Should return all program-registration-attributes on Get FSP instruction with Excel-FSP when "columnsToExport" is not set', async () => {
      // Arrange
      const programAttributeColumns =
        programTest.programRegistrationAttributes.map((pa) => pa.name);
      programAttributeColumns.concat(['amount']);

      const fspConfigurations = await getProgramFspConfigurations({
        programId: programIdWesteros,
        accessToken,
      });

      for (const fspConfiguration of fspConfigurations.body) {
        await deleteProgramFspConfigurationProperty({
          programId: programIdWesteros,
          configName: fspConfiguration.name,
          propertyName: FspConfigurationProperties.columnsToExport,
          accessToken,
        });
      }

      // Act
      const fspInstructionsResponse = await getFspInstructions(
        programIdWesteros,
        paymentIdWesteros,
        accessToken,
      );
      // Assert
      const programRegistrationAttributeNames =
        programTest.programRegistrationAttributes.map((pa) => pa.name);

      expect(fspInstructionsResponse.statusCode).toBe(HttpStatus.OK);

      const fspInstructions = fspInstructionsResponse.body;

      const expectedKeysPerRow = programRegistrationAttributeNames.concat([
        'amount',
        GenericRegistrationAttributes.referenceId,
      ]);
      for (const dataPerFsp of fspInstructions) {
        for (const row of dataPerFsp.data) {
          const fspInstructionColumns = Object.keys(row);
          expect(fspInstructionColumns.sort()).toEqual(
            expectedKeysPerRow.sort(),
          );
        }
      }
    });
  });

  describe('Error cases', () => {
    // None of these do write operations, so it's fine to not reset the DB only once
    beforeAll(async () => {
      await seedPrograms();
    });
    it("Sould throw an error if the payment doesn't exist", async () => {
      // Arrange
      const nonExistingPaymentId = 9999;

      // Act
      const fspInstructionsResponse = await getFspInstructions(
        programIdWesteros,
        nonExistingPaymentId,
        accessToken,
      );

      // Assert
      expect(fspInstructionsResponse.statusCode).toBe(HttpStatus.NOT_FOUND);
      expect(fspInstructionsResponse.body).toMatchSnapshot();
    });

    it('Should throw an error if program does not have Excel FSP configured', async () => {
      // Arrange

      // Act
      const fspInstructionsResponse = await getFspInstructions(
        programIdCbe,
        paymentIdCbe,
        accessToken,
      );

      // Assert
      expect(fspInstructionsResponse.statusCode).toBe(HttpStatus.NOT_FOUND);
      expect(fspInstructionsResponse.body).toMatchSnapshot();
    });

    it('Should throw an error if no transactions are found for the payment related to an excel FSP', async () => {
      // Act
      const fspInstructionsResponse = await getFspInstructions(
        programIdWesteros,
        voucherPaymentIdWesteros,
        accessToken,
      );

      // Assert
      expect(fspInstructionsResponse.statusCode).toBe(HttpStatus.NOT_FOUND);
      expect(fspInstructionsResponse.body).toMatchSnapshot();
    });
  });
});
