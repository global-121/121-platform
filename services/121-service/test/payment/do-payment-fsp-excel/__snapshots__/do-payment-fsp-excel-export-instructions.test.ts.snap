// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Do payment with Excel FSP Error cases Should throw an error if no transactions are found for the payment related to an excel FSP 1`] = `
{
  "message": "No transactions found for this payment with FSPs that require to download payment instructions.",
  "statusCode": 404,
}
`;

exports[`Do payment with Excel FSP Error cases Should throw an error if program does not have Excel FSP configured 1`] = `
{
  "message": "No program FSP configuration with Excel FSP found for this program",
  "statusCode": 404,
}
`;

exports[`Do payment with Excel FSP Error cases Sould throw an error if the payment doesn't exist 1`] = `
{
  "message": "Payment with ID 9999 not found in program 2",
  "statusCode": 404,
}
`;

exports[`Do payment with Excel FSP Export FSP instructions successfully Should return all program-registration-attributes on Get FSP instruction with Excel-FSP when "columnsToExport" is not set 1`] = `
[
  {
    "data": [
      {
        "accountId": null,
        "amount": 10,
        "date": null,
        "dob": "31-08-1990",
        "dragon": "0",
        "fixedChoice": null,
        "fullName": "Jaime Lannister",
        "healthArea": "",
        "house": "lannister",
        "knowsNothing": "false",
        "motto": "A lanister always pays his debts",
        "openAnswer": "",
        "personalId": "",
        "phoneNumber": "14155235556",
        "referenceId": "westeros987654322",
        "whatsappPhoneNumber": "14155235555",
      },
    ],
    "fileNamePrefix": "gringotts",
  },
  {
    "data": [
      {
        "accountId": null,
        "amount": 20,
        "date": null,
        "dob": "31-08-1990",
        "dragon": "1",
        "fixedChoice": null,
        "fullName": "John Snow",
        "healthArea": "",
        "house": "stark",
        "knowsNothing": "true",
        "motto": "Winter is coming",
        "openAnswer": "",
        "personalId": "",
        "phoneNumber": "14155235554",
        "referenceId": "westeros123456789",
        "whatsappPhoneNumber": "14155235554",
      },
      {
        "accountId": null,
        "amount": 10,
        "date": null,
        "dob": "31-08-1990",
        "dragon": "0",
        "fixedChoice": null,
        "fullName": "Arya Stark",
        "healthArea": "",
        "house": "stark",
        "knowsNothing": "false",
        "motto": "A girl has no name",
        "openAnswer": "",
        "personalId": "",
        "phoneNumber": "14155235555",
        "referenceId": "westeros987654321",
        "whatsappPhoneNumber": "14155235555",
      },
    ],
    "fileNamePrefix": "ironBank",
  },
]
`;
